### DISCUSSION TITLE
Have watchers become asynchronous in v3?

### DISCUSSION THREAD

**Discussion by northern-mindset** [alpinejs/alpine#2699] [1. Help] [answered]
Upgrading a project from v2 to v3.

In v2 I could set one property to one value and then to another value in the same event handler, and this triggered the watcher callback two times with respective values.

Now it seems that the watcher callbacks are

1. deferred until all events are processed (`nextTick`?)
2. and are squashed into only one callback with the latest value.

Is it by design or by mistake?

**Comment by SimoTod** [comment, marked as answer]
it's by design. Watchers use the Vue reactivity effects and the callback is deferred otherwise it would pick up variables used in the callback itself as a trigger (See https://github.com/alpinejs/alpine/blob/main/packages/alpinejs/src/magics/%24watch.js#L16).

**Reply by northern-mindset** [reply]
Thank you for fast reply.

> otherwise it would pick up variables used in the callback itself as a trigger

Do you mean here that if the watch callback changes any data properties, this would trigger it again in a loop?

And is it true that in v2 it worked the other way?

**Reply by SimoTod** [reply]
I don't think but I didn't write the code so I'm not 100% sure. I think the comment means that when you reference another x-data variable in your callback if it wasn't deferred, the watcher would also trigger when the other variable changes (because it's the way effects work, see https://alpinejs.dev/advanced/reactivity#alpine-effect).

V2 was using a completely different implementation, not sure how it worked for your specific use case. In general, Js is single-threaded so if you change the same variable 2 times in the same callback, there's no time for other statements to run any code while the first value is set until you don't force the callback to release the flow (setTimeout, queueMicrotask, requestAnimationFrame or something like those) so probably your watcher was running twice but after the variable had already the second value set.
